src/shape.rs

```diff
@@ -23,53 +23,37 @@ pub struct Shape {
 }
 
 impl Shape {
-    /// Common function to create a new `Shape` with calculating inner statistics.
+    /// Creates a new n-dimensional shape.
     ///
     /// # Arguments
     ///
-    /// * `num_dimensions` - Number of dimensions in the new shape.
-    /// * `dimensions` - Number of elements for each dimension.
+    /// * `dimensions` - Number of elements for each axis.
     ///
     /// # Returns
     ///
     /// A new `Shape` object.
     ///
     /// # Panics
     ///
-    /// `num_dimensions` exceeds MAX_NUM_DIMENSIONS.
-    fn new_inner(num_dimensions: usize, dimensions: [usize; MAX_NUM_DIMENSIONS]) -> Self {
+    /// The const parameter `N` is larger than `MAX_NUM_DIMENSIONS`.
+    pub fn new<const N: usize>(dimensions: [usize; N]) -> Self {
+        let (dimensions, num_elements) = {
+            let mut buffer = [0usize; MAX_NUM_DIMENSIONS];
+            let mut size = 1usize;
+            for i in 0..N {
+                let d = dimensions[i];
+                buffer[i] = d;
+                size *= d;
+            }
```

Comment:
Optimization is more expected using `zip` instead of indices.
`[i]` inserts boundary checking every iteration.
```suggestion
            assert!(dimensions.len() <= MAX_NUM_DIMENSIONS, "error message");
            for (b, d) in buffer.iter_mut().zip(dimensions) {
                *b = d;
                size *= d;
            }
```

Reply:
As the function takes size-known array, optimizer can avoid boundary checking.

Reply:
a simplified working example:
<img width="914" alt="Untitled" src="https://user-images.githubusercontent.com/1023695/155529129-c252b759-73ea-4c87-b93a-a7650e54d116.png">

Reply:
Add comment to clarify this point. e258039

Reply:
for the same reason I realized that it is reasonable to add assertion you suggested: it consists of only static values and the optimizer can avoid the assertion if necessary.

Reply:
Since maybe applying the zipped iterator results in a similar codegen, I switched the implementation to use the iterator.
I also added `from_slice()` function.
acfd2d8

Reply:
> As the function takes size-known array, optimizer can avoid boundary checking.

I know that. Unnecessary boundary checks will be removed in optimization.
However, why do you use useless indices?

Reply:
> useless indices?

What is "useless indices" in this context? If you mean `i`, it is only the difference of styles, or just a compiler hacks (compilers usually implements an optimization of trivial indexing which results in more efficient code than we expected).

Anyway I changed the old implementation because the codegen does eventually not affect.

Reply:
Examples: both works similarly.
Personally I like indexing, because it is much simpler than zipped iterators.
<img width="1297" alt="Untitled" src="https://user-images.githubusercontent.com/1023695/155545464-c61f33da-d6e8-48a4-9083-8fd111d5a9bd.png">